drivers/azure/azure.go

```diff
@@ -0,0 +1,455 @@
+		func() error { return c.CreateStorageAccount(d.ctx, d.ResourceGroup, d.Location, defaultStorageType) },
+		func() error { return d.generateSSHKey(d.ctx) },
+		func() error {
+			return c.CreateVirtualMachine(d.ResourceGroup, d.naming().VM(), d.Location, d.Size, d.ctx.AvailabilitySetID,
+				d.ctx.NetworkInterfaceID, d.BaseDriver.SSHUser, d.ctx.SSHPublicKey, d.Image, d.ctx.StorageAccount)
+		},
+	}
+	for _, f := range steps {
+		if err := f(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Remove deletes the virtual machine and resources associated to it.
+func (d *Driver) Remove() error {
+	// NOTE(ahmetalpbalkan):
+	//   - remove attemps are best effort and if a resource is already gone, we
+	//     continue removing other resources instead of failing.
+	//   - we can probably do a lot of parallelization here but a sequential
+	//     logic works fine too. If we were to detach the NIC from the VM and
+	//     then delete the VM, this could enable some parallelization.
+
+	log.Info("NOTICE: Please check Azure portal/CLI to make sure you have no leftover resources to avoid unexpected charges.")
+	c, err := d.newAzureClient()
+	if err != nil {
+		return err
+	}
+	for _, f := range []func() error{
+		func() error { return c.DeleteVirtualMachineIfExists(d.ResourceGroup, d.naming().VM()) },
+		func() error { return c.DeleteNetworkInterfaceIfExists(d.ResourceGroup, d.naming().NIC()) },
+		func() error { return c.DeletePublicIPAddressIfExists(d.ResourceGroup, d.naming().IP()) },
+		func() error { return c.DeleteNetworkSecurityGroupIfExists(d.ResourceGroup, d.naming().NSG()) },
+		func() error { return c.CleanupAvailabilitySetIfExists(d.ResourceGroup, d.AvailabilitySet) },
+		func() error { return c.CleanupSubnetIfExists(d.ResourceGroup, d.VirtualNetwork, d.SubnetName) },
+		func() error { return c.CleanupVirtualNetworkIfExists(d.ResourceGroup, d.VirtualNetwork) },
+	} {
+		if err := f(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// GetIP returns public IP address or hostname of the machine instance.
+func (d *Driver) GetIP() (string, error) {
+	if d.resolvedIP == "" {
+		ip, err := d.ipAddress()
+		if err != nil {
+			return "", err
+		}
+		d.resolvedIP = ip
+	}
+	log.Debugf("Machine IP address resolved to: %s", d.resolvedIP)
+	return d.resolvedIP, nil
+}
+
+// GetSSHHostname returns an IP address or hostname for the machine instance.
+func (d *Driver) GetSSHHostname() (string, error) {
+	return d.GetIP()
+}
+
+// GetURL returns a socket address to connect to Docker engine of the machine
+// instance.
+func (d *Driver) GetURL() (string, error) {
+	if err := drivers.MustBeRunning(d); err != nil {
+		return "", err
+	}
+
+	// NOTE (ahmetalpbalkan) I noticed that this is not used until machine is
+	// actually created and provisioned. By then GetIP() should be returning
+	// a non-empty IP address as the VM is already allocated and connected to.
+	ip, err := d.GetIP()
+	if err != nil {
+		return "", err
+	}
+	u := (&url.URL{
```

Comment:
@nathanleclaire hehe we can have a long argument about this. I'm very defensive about repeating the logic of stdlibs. For instance, people using `:` character to split IP/port are essentially going to be broken with IPv6. If should've used net.SplitHostPort, they'd be fine. So I'm basically using the logic in net/url to be on the safe side. :)